Fix architecture check on TemplateGen against the archs supported by the codestream

marcosps opened the pull request:
The original check was working by change. Each codestream has a set of supported architectures, like -RT ones being supported only on x86_64 for now. The current check was checking the supported archs on a codestream with the architectures that was affected on the project level, which is wrong.

The change checks the affected architectures on the codestream against the architectures that the codestream supports. The original tests also worked with this change, being even more clear that it worked by chance at the time.

marcosps committed:
templ: Check get_default_archs with cs.archs

The original check was working by change. Each codestream has a set
of supported architectures, like -RT ones being supported only on x86_64
for now. The current check was checking the supported archs on a
codestream with the architectures that was affected on the project level,
which is wrong.

The change checks the affected architectures on the codestream against
the architectures that the codestream supports. The original tests also
worked with this change, being even more clear that it worked by chance
at the time.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
ci-tests: Decrease the score

Somehow it's now failing with a recent commit, with no aparent reason.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps committed:
plugins: setup: Run __setup_check_syms for all symbols of an object

In my tests, before this patch, with a livepatch that affects two
files on a codestrams, it took 35 minutes. Now it takes around 10 minutes,
just by avoiding to read all the symbols more than once for the same
object.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps commented:
@fgyanz @vmezzela I just added a new commit to help reducing the time spent on setup. The next improvement should be by parallelizing it by by the same number of jobs that we currently use for extraction. I can think about it in the next days...

@fgyanz would you mind trying this PR on CI to check if this helps to reduce the slowness that you found there?

marcosps committed:
plugins: setup: Change the checks for missing symbol

Up until now, whenever a symbol can't be found on an arch a warning was
shown. This isn't really a problem anymore since klp-ccp can pick the
callers of inline functions automatically, so it just adds more concern
to livepatch developers.

Change the procedure to check if the symbol is missing on all kernels,
which means that the same inline decision was made for all of them. But
if the symbol is present in one arch and not in another, this means that
maybe we can have issues, so print the warning as before.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

fgyanz commented on klpbuild/plugins/setup.py:
```diff
@@ -191,7 +193,19 @@ def setup_project_files(lp_name, codestreams):
+
+                slist.setdefault(mod, {})
+
+            data = slist.get(mod, {})
```
Just nitpicking, so feel free to skip this comment. 
I'd propose renaming this variable to something more specific. "data" could mean anything in this context. I'd also propose changing "slist" to "syms" if it's a list of symbols. I think it would be more readable that way.

marcosps replied:
There is already a variable called `syms` in the same function =/

vmezzela replied:
I agree, I would change the name of the dict and make it clear what is it meant for. Something like `syms_to_be_checked`.

marcosps replied:
agreed!

fgyanz commented on .github/workflows/ci-tests.yml:
```diff
@@ -27,7 +27,7 @@ jobs:
       run: >
         tox -e lint --
-        --fail-under 9.85
+        --fail-under 9.83
```
I'm not really sure this is the proper fix. I mean, it works as a workaround now, but will we need to keep doing it every now and then? Maybe we should rethink the approach for a more solid solution.

marcosps replied:
Yeah, but unfortunately pylint shows really shows what decreased last time, and when I try to run pylint locally it still shows the score as 9.86, so I'm not sure that changed...

fgyanz reviewed (approved):
Just two comments! Not mandatory to fix them, feel free to decide what to do. Thanks for working on the PR so fast! LGTM!

vmezzela commented on klpbuild/plugins/setup.py:
```diff
@@ -244,6 +244,13 @@ def __setup_check_mod(cs, mod, arch):
+    # interesting to check in other architectures.
+    if arch == utils.ARCH:
+        return
+
```
I am not completely sure this is the right way to go. If a symbol is not present in `x86_64` and on the other archs, then klp-build would still log the warning for the other archs even though the code extracted by klp-ccp would (hopefully) be fine assuming the inlining decision is the same across all the archs. Hence, we'd have the warning in a totally useless case.

Consider the scenario where a given symbol is inlined only in `x86_64` and not on the other archs, there is no way for us to find out that on the other archs we need to patch those syms as well. Hence, here we'd miss the warning in a totally useful case.

So, rather that dropping the warning altogether for `x86_64`, I'd propose to log it anyway. But amending the message so that the user knows that it's not necessarily an issue in case the symbol is missing.

fgyanz replied:
> I am not completely sure this is the right way to go. If a symbol is not present in x86_64 and on the other archs, then klp-build would still log the warning for the other archs even though the code extracted by klp-ccp would (hopefully) be fine assuming the inlining decision is the same across all the archs. Hence, we'd have the warning in a totally useless case.

right, good catch. I'd avoid printing the warning if all the archs are missing the symbol. Most of the times this is case, which means the symbols is inlined.

marcosps replied:
Ok, let me change the patch then!

vmezzela commented on klpbuild/plugins/setup.py:
```diff
@@ -191,7 +193,19 @@ def setup_project_files(lp_name, codestreams):
                 __setup_check_mod(cs, mod, arch)
-                __setup_check_syms(cs, mod, syms, arch)
+
+                slist.setdefault(mod, {})
```
I think that this line can be remove altogether: it's redundant with the `slist.get(mod, {})` ?

marcosps replied:
I believe so. I changed this code so many times that somehow it was needed at some point...

marcosps commented:
@fgyanz @vmezzela so I changed the check for showing the warning for missing symbols, please take a look!

fgyanz committed:
plugins/extract: Apply the patches in numerical order

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz commented:
Changes make sense! LGTM!

vmezzela commented on klpbuild/klplib/codestream.py:
```diff
@@ -194,6 +194,7 @@ def set_configs(self, configs):
 
     def set_archs(self, archs):
         self.archs = list(set(archs) & set(self.get_default_archs()))
+        self.archs.sort()
```
This commit made me realize that `__is_check_enabled()` relies on the two lists  `cs.archs != cs.get_default_archs()` being the same also in their order, as you mentioned in the commit message.

If that's the case, can we readapt the code to use a `set()` for the archs field in codestreams?

So we avoid similar bugs in the future around this code. Feel free also to add a patch on top of the current branch instead of rebasing - otherwise I can also address this in my PR.

vmezzela commented on klpbuild/plugins/setup.py:
```diff
@@ -242,9 +241,40 @@ def __setup_check_mod(cs, mod, arch):
+    sym_archs = {}
+    for larch, lsyms in mod_syms.items():
+        for sym in lsyms:
+            sym_archs.setdefault(sym, [])
```
If you decide to apply the change suggested in the previous comment, this would became `sym_archs.setdefault(sym, set())`, and you can drop the `larchs.sort()` below so we no longer have to rely on the list being sorted (we might forget about that in the future).

vmezzela reviewed (changes requested):
Hi @marcosps, the PR looks good to me. I left a proposed change anyway if you want to address it. I believe it'd make the code less prone to future bugs around the way we use `cs.archs`. Otherwise I apply it in my PR if you like the proposal.

marcosps committed:
klplib: codestream: Convert cs.archs into a set

This way we don't need to sort the elements.

Suggested-by: Vincenzo MEZELLA <vincenzo.mezzela@suse.com>
Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>

marcosps commented:
> Hi @marcosps, the PR looks good to me. I left a proposed change anyway if you want to address it. I believe it'd make the code less prone to future bugs around the way we use `cs.archs`. Otherwise I apply it in my PR if you like the proposal.

Done! Thanks for the suggestion! Adapting the code to a set required some more changes, but in the end it was worth it!

Also, in the symbol check, I had to keep it it as a list, and then later on convert it as a set, otherwise we can't append the archs into set. I believe that I addressed all your comments!
